File: client/golang/prsim/locker.go
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package prsim

import (
	"context"
	"time"
)

var ILocker = (*Locker)(nil)

// Locker spi
// @MPI(macro.MeshMPI)
// @SPI(macro.MeshSPI)
type Locker interface {

	// Lock create write lock.
	// @MPI("mesh.locker.w.lock")
	Lock(ctx context.Context, rid string, timeout time.Duration) (bool, error)

	// Unlock release write lock.
	// @MPI("mesh.locker.w.unlock")
	Unlock(ctx context.Context, rid string) error

	// ReadLock create read lock.
	// @MPI("mesh.locker.r.lock")
	ReadLock(ctx context.Context, rid string, timeout time.Duration) (bool, error)

	// ReadUnlock release read lock.
	// @MPI("mesh.locker.r.unlock")
	ReadUnlock(ctx context.Context, rid string) error
}

File: client/javascript/src/prsim/locker.ts
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

import {spi} from "@/macro/spi";
import {mpi} from "@/macro/mpi";
import type {Context} from "@/prsim/context";
import {Status} from "@/cause";
import {index} from "@/macro";

@spi("mesh")
export abstract class Locker {

    /**
     * Lock create write lock.
     */
    @mpi("mesh.locker.w.lock", Boolean)
    lock(ctx: Context, @index(0, 'rid') rid: string, @index(1, 'timeout') timeout: number): Promise<boolean> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * Unlock release write lock.
     */
    @mpi("mesh.locker.w.unlock")
    unlock(ctx: Context, @index(0, 'rid') rid: string): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * ReadLock create read lock.
     */
    @mpi("mesh.locker.r.lock", Boolean)
    readLock(ctx: Context, @index(0, 'rid') rid: string, @index(1, 'timeout') timeout: number): Promise<boolean> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    /**
     * ReadUnlock release read lock.
     */
    @mpi("mesh.locker.r.unlock")
    readUnlock(ctx: Context, @index(0, 'rid') rid: string): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }
}